JavaScript/JavaScriptSDK/ajax/ajax.ts

```diff
@@ -291,211 +58,284 @@ module Microsoft.ApplicationInsights {
+                        var ajaxData = new ajaxRecord();
+                        ajaxData.method = method;
+                        ajaxData.requestUrl = url;
+                        ajaxData.requestSize += url.length;
                                 // If not set async defaults to true 
-                                self.ajaxData.async = extensions.IsNullOrUndefined(async) ? true : async;
+                        ajaxData.async = extensions.IsNullOrUndefined(async) ? true : async;
+                        this.ajaxData = ajaxData;
 
-                            });
+                        ajaxMonitorInstance.attachToOnReadyStateChange(this);
+                    }
+                } catch (e) {
+                    _InternalLogging.throwInternalNonUserActionable(
+                        LoggingSeverity.CRITICAL,
+                        "Failed to monitor XMLHttpRequest.open, monitoring data for this ajax call may be incorrect: "
+                        + Microsoft.ApplicationInsights.Util.dump(e));
                         }
 
                         return originalOpen.apply(this, arguments);
                     };
                 }
 
-                function interceptSend() {
+        private instrumentSetRequestHeader() {
+            var originalSetRequestHeader = XMLHttpRequest.prototype.setRequestHeader;
+            var ajaxMonitorInstance = this;
+            XMLHttpRequest.prototype.setRequestHeader = function (name, value) {
+                try {
+                    if (ajaxMonitorInstance.isMonitoredInstance(this)) {                    
+                        // 2 is the length of ": " which is added to each header
+                        this.ajaxData.requestSize += stringUtils.GetLength(name) + stringUtils.GetLength(value) + 2;
+                    }
+                } catch (e) {
+                    _InternalLogging.throwInternalNonUserActionable(
+                        LoggingSeverity.CRITICAL,
+                        "Failed to monitor XMLHttpRequest.setRequestHeader, monitoring data for this ajax call may be incorrect: "
+                        + Microsoft.ApplicationInsights.Util.dump(e));
+                }
+
+                return originalSetRequestHeader.apply(this, arguments);
+            };
+        }
+
+        private instrumentSend() {
                     var originalSend = XMLHttpRequest.prototype.send;
+            var ajaxMonitorInstance = this;
                     XMLHttpRequest.prototype.send = function (content) {
-                        sendPrefixInterceptor.call(this, content);
+                try {
+                    ajaxMonitorInstance.sendPrefixInstrumentor(this, content);
+                } catch (e) {
+                    _InternalLogging.throwInternalNonUserActionable(
+                        LoggingSeverity.CRITICAL,
+                        "Failed to monitor XMLHttpRequest.send, monitoring data for this ajax call may be incorrect: "
```

Comment:
Is it guaranteed that the original ajax call will still be made? Is there is no guarantee can we do emergency shut off of all further ajax monitoring calls? Is there is guarantee - let's say about it in the error message.

Also - report url or some other identifier of the original ajax call